### PR TITLE
Fixed #27654 -- Propagated alters_data attribute to callables overridden in subclasses.

### PR DESCRIPTION
**Pull request by lightdiscord**
Ticket: https://code.djangoproject.com/ticket/27654

Does not fix the issue for 3rd party libraries but tries to fix this at least here.

#### Progress

Methods where alters_data is set to True

* [x] class FieldFile
	* [x] class ImageFieldFile
		* [x] `delete` overriden but no `alters_data` attribute
	* [x] class FakeFieldFile

```
django/db/models/fields/files.py:    open.alters_data = True
django/db/models/fields/files.py:    save.alters_data = True
django/db/models/fields/files.py:    delete.alters_data = True
```

* [x] class RelatedManager (hard to follow usage)

```
django/db/models/fields/related_descriptors.py:        add.alters_data = True
django/db/models/fields/related_descriptors.py:        create.alters_data = True
django/db/models/fields/related_descriptors.py:        get_or_create.alters_data = True
django/db/models/fields/related_descriptors.py:        update_or_create.alters_data = True
django/db/models/fields/related_descriptors.py:            remove.alters_data = True
django/db/models/fields/related_descriptors.py:            clear.alters_data = True
django/db/models/fields/related_descriptors.py:            _clear.alters_data = True
django/db/models/fields/related_descriptors.py:        set.alters_data = True
```

* [x] class ManyRelatedManager (hard to follow usage)

```
django/db/models/fields/related_descriptors.py:        add.alters_data = True
django/db/models/fields/related_descriptors.py:        remove.alters_data = True
django/db/models/fields/related_descriptors.py:        clear.alters_data = True
django/db/models/fields/related_descriptors.py:        set.alters_data = True
django/db/models/fields/related_descriptors.py:        create.alters_data = True
django/db/models/fields/related_descriptors.py:        get_or_create.alters_data = True
django/db/models/fields/related_descriptors.py:        update_or_create.alters_data = True
```

* [x] class Model
	* [x] class Site
	* [x] class AbstractBaseUser
		* [x] `save` overriden but no `alters_data` attribute
		* [x] class AbstractUser
			* [x] class User
	* [x] class Permission
	* [x] class Group
	* [x] class PermissionsMixin
		* [x] class AbstractUser (already checked)
	* [x] class PostGISGeometryColumns
	* [x] class PostGISSpatialRefSys
	* [x] class SpatialiteGeometryColumns
	* [x] class SpatialiteSpatialRefSys
	* [x] class OracleGeometryColumns
	* [x] class OracleSpatialRefSys
	* [x] class FlatPage
	* [x] class Redirect
	* [x] class LogEntry
	* [x] class ContentType
	* [x] class AbstractBaseSession
		* [x] class Session
	* Classes in `tests/`, ignored.

```
django/db/models/base.py:    save.alters_data = True
django/db/models/base.py:    save_base.alters_data = True
django/db/models/base.py:    delete.alters_data = True
```

* [x] class QuerySet
	* [x] class Manager (does not directly inherit from QuerySet but checks if some of the methods that alters data are simply re-exposed)
		* [x] class EmptyManager
		* [x] class CurrentSiteManager
		* [x] class SiteManager
		* [x] class BaseUserManager
			* [x] class UserManager
		* [x] class PermissionManager
		* [x] class GroupManager
		* [x] class LogEntryManager
		* [x] class ContentTypeManager
		* [x] class BaseSessionManager
			* [x] class SessionManager

```
django/db/models/query.py:    bulk_update.alters_data = True
django/db/models/query.py:    abulk_update.alters_data = True
django/db/models/query.py:    delete.alters_data = True
django/db/models/query.py:    adelete.alters_data = True
django/db/models/query.py:    _raw_delete.alters_data = True
django/db/models/query.py:    update.alters_data = True
django/db/models/query.py:    aupdate.alters_data = True
django/db/models/query.py:    _update.alters_data = True
django/db/models/query.py:    _insert.alters_data = True
```

* [x] class GenericRelatedObjectManager (hard to follow usage)

```
django/contrib/contenttypes/fields.py:        add.alters_data = True
django/contrib/contenttypes/fields.py:        remove.alters_data = True
django/contrib/contenttypes/fields.py:        clear.alters_data = True
django/contrib/contenttypes/fields.py:        _clear.alters_data = True
django/contrib/contenttypes/fields.py:        set.alters_data = True
django/contrib/contenttypes/fields.py:        create.alters_data = True
django/contrib/contenttypes/fields.py:        get_or_create.alters_data = True
django/contrib/contenttypes/fields.py:        update_or_create.alters_data = True
```

* [x] class BaseModelForm
	* [x] class ModelForm
		* [x] class UserCreationForm
			* [x] `save` overriden but no `alters_data` attribute
		* [x] class UserChangeForm
		* [x] class FlatpageForm
		* Some classes in `tests/`, ignored.

```
django/forms/models.py:    save.alters_data = True
```

* [x] class BaseModelFormSet
	* [x] class BaseGenericInlineFormSet
	* [x] class BaseInlineFormSet
	* [x] class BasePersonModelFormSet

```
django/forms/models.py:    save.alters_data = True
```

* [x] class Doodad

```
tests/template_tests/test_callables.py:            alters_data = True
```

* [x] class Doodad

```
tests/template_tests/test_callables.py:            alters_data = True
```

### CODE REVIEW

**Review by carltongibson — COMMENTED** [review]
Thanks for this @lightdiscord — I'd be inclined to take this as it is, but I'll ask @adamchainz and @shaib if they have any input, because they both mentioned _other ideas™_ in discussion. 

🎁 

(We should likely add a regression test too 🤔)

**Comment by shaib** [comment]
Hi @lightdiscord , 

Thanks for this -- while I think the changes, per se, are not the correct long-term fix, the work you've put into finding all the uses is very valuable and much appreciated (and surely represents a considerable effort). And certainly, if we start getting bogged-down in discussions, this is better than nothing.

What I think should be the proper solution is something along the lines of (untested code):

```python
def preserve_alters_data(mro, namespace):
    for name, obj in namespace.items():
        if callable(obj) and not hasattr(obj, 'alters_data'):
            for base in mro:
                base_obj = getattr(base, name, None)
                if base_obj:
                    if hasattr(base_obj, 'alters_data'):
                        obj.alters_data = base_obj.alters_data
                    break  # Whether it has alters_data or not, the attr was found
```

and this should be called on class creation, for any of the classes you've listed above.

Doing this would make the specific fixes here redundant, and should solve the issue for user classes as well, but of course it is a deeper change.

**Comment by lightdiscord** [comment]
With the three dynamically generated classes it's easy to lose track of usages.

About other ideas for this problem, I've mentioned in the mailing-list a solution where it should be possible to traverse each base class to search a method with the same name and check for the attribute on it. Unfortunately, the [PoC](https://gist.github.com/lightdiscord/fd03fd6bf1a775f23f3981f07f15a0bf) I sent does not cover a case which is used that fails (and I didn't find a solution yet).

A downside with this method is if the function that alters data is re-exposed with another name it also doesn't catch it. Another thing I thought of (but do not have a PoC yet) is to have a thread local, a contextvar or whatever that says if a template is currently rendering and each function that alters data will check this and decides to throw if that's the case. (The problem being that function that may not have thrown exception before will now do).

> We should likely add a regression test too :thinking: 

The regression test is only a check to see if the alters_data attribute is present on the `class User` for example? Or something else?

**Comment by shaib** [comment]
I (natuarally, in view of my own suggested solution) like the general approach in your PoC; the thing I don't like about it is that it would be doing this relatively expensive lookup in rendering, and as presented, repeat it every time. The approach I suggested here (and also what @adamchainz suggested before) moves this computation to class creation time, once.

**Comment by carltongibson** [comment]
> ...regression test... 

Assuming @shaib's approach, we'd have a model class overriding ≈`save()` but not defining `alters_data` and assert that it was present on the final class. (Or such...)

File: django/template/utils.py [hunk 1]
```diff
@@ -109,3 +110,19 @@ def get_app_template_dirs(dirname):
     ]
     # Immutable return value because it will be cached and shared by callers.
     return tuple(template_dirs)
+
+
+class AltersData:
+    def __init_subclass__(cls, **kwargs):
+        mro = inspect.getmro(cls)[1:]
```
Comment (adamchainz) [inline django/template/utils.py:117]: Just use `cls.__mro__`, `inspect.getmro` seems to be useless

https://github.com/python/cpython/blob/8af04cdef202364541540ed67e204b71e2e759d0/Lib/inspect.py#L729-L731

Also I think you really only need to loop over `cls.__bases__`? https://docs.python.org/3.10/library/stdtypes.html#class.__bases__ . Because if `getattr` finds a function on a class, it exists either in that class or in its ancestors, so that covers the ancestors in the MRO.

File: django/contrib/contenttypes/fields.py [hunk 2]
```diff
@@ -560,7 +561,7 @@ def create_generic_related_manager(superclass, rel):
     specific to generic relations.
     """
 
-    class GenericRelatedObjectManager(superclass):
+    class GenericRelatedObjectManager(superclass, AltersData):
```
[the remark this answers is not in the document]
Comment (lightdiscord) [reply django/contrib/contenttypes/fields.py:564]: I don't understand why it shouldn't be necessary if `superclass` is always another manger?
I looked through the code and also logged the `superclass` for a lot of tests and yes it always seems to be a `db.models.Manager` (or something that inherits from it). But `AltersData` is not an (in)direct base class of `Manager`.

[the remark this answers is not in the document]
Comment (adamchainz) [reply django/contrib/contenttypes/fields.py:564]: Apologies, I thought you had added `AltersData` to `Manager`. But perhaps we should do that?

[the remark this answers is not in the document]
Comment (lightdiscord) [reply django/contrib/contenttypes/fields.py:564]: The benefit of having `AltersData` directly on the class that sets some `alters_data` is that it's easy to check if the base class `AltersData` is missing or not.
In the tests I did on the codebase I didn't see a case where `GenericRelatedObjectManager` did not inherit from something else than django's Manager, but I don't know about someone's else codebase.

**Comment by shaib** [comment]
Thanks for all the work -- this looks good to me, with one exception (a point I think has been raised before): `class AltersData` should not sit in `django.template.utils`. Even if the only use currently in Django is in the template engines, this is a models concept which can be used with no relation to templates; and further, nothing in django.db.models` should depend on anything from `django.templates`.

I'd pick `django.db.models.utils`, probably.

File: django/db/models/utils.py [hunk 3]
```diff
@@ -50,3 +50,17 @@ def __reduce__(self):
         (namedtuple("Row", names),),
         {"__reduce__": __reduce__, "__slots__": ()},
     )
+
+
+class AltersData:
```
Comment (shaib) [inline django/db/models/utils.py:55]:
```suggestion
class AltersData:
    """Make subclasses preserve the ``alters_data`` attribute on overridden methods."""
```

**Review by shaib — APPROVED** [review]
In light of the `__mro__` vs. `__bases__` issue we went through, I might also add some tests with multiple and/or multi-level inheritance; other than that, LGTM.

**Comment by lightdiscord** [comment]
Hi, what's the next step to get this merged?

**Comment by lightdiscord** [comment]
Hi, any blocking step remaining before merging this branch? @carltongibson and @shaib

File: tests/template_tests/test_callables.py [hunk 4]
```diff
@@ -65,21 +65,38 @@ def __call__(self):
...
-        self.assertEqual(child.my_method.alters_data, True)
+            def other_method(self):
+                return 101
+
+        class GrandChild(Child):
+            pass
+
+        element = GrandChild()
+        self.assertEqual(element.my_method.alters_data, True)
+        self.assertEqual(element.other_method.alters_data, True)
```
Comment (shaib) [inline tests/template_tests/test_callables.py:95]:
```suggestion
        self.assertTrue(element.my_method.alters_data)
        self.assertTrue(element.other_method.alters_data)
```

File: tests/template_tests/test_callables.py [hunk 5]
```diff
@@ -65,21 +65,38 @@ def __call__(self):
...
 
-        c = Context({"child": child})
-        t = self.engine.from_string("{{ child.my_method }}")
+        c = Context({"element": element})
+        t = self.engine.from_string("{{ element.my_method }}")
+        t = self.engine.from_string("{{ element.other_method }}")
```
Comment (shaib) [inline tests/template_tests/test_callables.py:99]: This looks like an error...

File: tests/template_tests/test_callables.py [hunk 6]
```diff
@@ -63,6 +64,41 @@ def __call__(self):
...
+
+            def other_method(self):
+                return 101
+
+        class GrandChild(Child):
+            pass
+
+        element = GrandChild()
+        self.assertEqual(element.my_method.alters_data, True)
+        self.assertEqual(element.other_method.alters_data, True)
```
Comment (felixxm) [inline tests/template_tests/test_callables.py:95]:
```suggestion
        self.assertIs(element.my_method.alters_data, True)
        self.assertIs(element.other_method.alters_data, True)
```

File: django/contrib/contenttypes/fields.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamchainz) [reply django/contrib/contenttypes/fields.py:564]: I don’t think it's possible for the class to not inherit from `Manager`. But, you're right, this is best kept local to where `alters_data` is set, so we can stick with the current approach.

**Comment by felixxm** [comment]
@lightdiscord Thanks :+1: Welcome aboard :boat: 

I added extra assertions.

**Comment by adamchainz** [comment]
Thanks!

**Comment by shaib** [comment]
Thanks!
